src/extra/widgets/animbtn/lv_animbtn.c

```diff
@@ -0,0 +1,309 @@
+ */
+void lv_animbtn_set_state_desc(lv_obj_t * obj, lv_animbtn_state_t state, const lv_animbtn_state_desc_t desc)
+{
+    LV_ASSERT_OBJ(obj, MY_CLASS);
+
+    lv_animbtn_t * animbtn = (lv_animbtn_t *)obj;
+
+    animbtn->state_desc[state] = desc;
+    animbtn->state_desc[state].control |= 0x80; /*A non existant flag used to mark that the state was used*/
+    apply_state(obj);
+}
+
+void lv_animbtn_set_state(lv_obj_t * obj, lv_animbtn_state_t state)
+{
+    LV_ASSERT_OBJ(obj, MY_CLASS);
+
+    lv_state_t obj_state = LV_STATE_DEFAULT;
+    if(state == LV_ANIMBTN_STATE_PRESSED || state == LV_ANIMBTN_STATE_CHECKED_PRESSED) obj_state |= LV_STATE_PRESSED;
+    if(state == LV_ANIMBTN_STATE_DISABLED || state == LV_ANIMBTN_STATE_CHECKED_DISABLED) obj_state |= LV_STATE_DISABLED;
+    if(state == LV_ANIMBTN_STATE_CHECKED_DISABLED || state == LV_ANIMBTN_STATE_CHECKED_PRESSED ||
+       state == LV_ANIMBTN_STATE_CHECKED_RELEASED) {
+        obj_state |= LV_STATE_CHECKED;
+    }
+
+    lv_obj_clear_state(obj, LV_STATE_CHECKED | LV_STATE_PRESSED | LV_STATE_DISABLED);
+    lv_obj_add_state(obj, obj_state);
+
+    apply_state(obj);
+}
+
+/*=====================
+ * Getter functions
+ *====================*/
+
+/**
+ * Get the right image in a given state
+ * @param obj pointer to an image button object
+ * @param state the state where to get the image (from `lv_btn_state_t`) `
+ * @return pointer to the left image source (a C array or path to a file)
+ */
+const lv_animbtn_state_desc_t * lv_animbtn_get_state_desc(lv_obj_t * obj, lv_animbtn_state_t state)
+{
+    LV_ASSERT_OBJ(obj, MY_CLASS);
+    lv_animbtn_t * animbtn = (lv_animbtn_t *)obj;
+
+    return &animbtn->state_desc[state];
+}
+
+
+/**********************
+ *   STATIC FUNCTIONS
+ **********************/
+
+static void lv_animbtn_constructor(const lv_obj_class_t * class_p, lv_obj_t * obj)
+{
+    LV_UNUSED(class_p);
+    lv_animbtn_t * animbtn = (lv_animbtn_t *)obj;
+    /*Initialize the allocated 'ext'*/
+    lv_memset_00(animbtn->state_desc, sizeof(animbtn->state_desc));
+    animbtn->lottie = NULL;
+    animbtn->prev_state = _LV_ANIMBTN_STATE_NUM;
+
+    lv_obj_add_flag(obj, LV_OBJ_FLAG_CLICKABLE);
+    lv_obj_add_flag(obj, LV_OBJ_FLAG_CHECKABLE);
+}
+
+
+static void lv_animbtn_event(const lv_obj_class_t * class_p, lv_event_t * e)
+{
+    LV_UNUSED(class_p);
+
+    lv_res_t res = lv_obj_event_base(&lv_animbtn_class, e);
+    if(res != LV_RES_OK) return;
+
+    lv_event_code_t code = lv_event_get_code(e);
+    lv_obj_t * obj = lv_event_get_current_target(e);
+    lv_animbtn_t * animbtn = (lv_animbtn_t *)obj;
+    switch(code) {
+        case LV_EVENT_READY:
+            /*Should we loop?*/
+            loop_state(obj);
+            break;
+        case LV_EVENT_PRESSED:
+        case LV_EVENT_RELEASED:
+        case LV_EVENT_PRESS_LOST:
+            apply_state(obj);
+            break;
+        case LV_EVENT_DRAW_MAIN:
+            //draw_main(e);
+            lv_obj_invalidate(animbtn->lottie);
```

Comment:
Why is it needed? An invalidation during drawing can mess up things. 

Lottie animations should invalidate themself as needed.

Reply:
Indeed, I've just tested it and it's not required. Will remove it.